### PR TITLE
Add indexer binary and database

### PR DESCRIPTION
**Pull request by jkutner**
Introduces the following changes:

* Adds database support
* Adds database migrations
* Adds a Golang binary that performs indexing and upserting into database
* Adds an `indexer` and fetches the Git repo and denormalizes it

### CODE REVIEW

File: bin/indexer [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    }
+})();
+
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function buildIndex(indexPath) {
+    try {
+        let subProc = execa('./bin/index-buildpacks', [indexPath])
```
Comment (elbandito) [inline bin/indexer:34]: Hey Joe, I was looking at this late Friday, and I'm pretty sure we need a `await` in front of this.

Comment (jkutner) [reply bin/indexer:34]: I'm not sure, but I think the `await` is a few lines below. This way, it sets up the std/stderr pipes first.

Comment (elbandito) [reply bin/indexer:34]: ah, I missed that 👍🏼

Comment (elbandito) [reply bin/indexer:34]: Yup, that's fine.   I must have been looking at an older commit.  This should be fine.

File: db/migrations/20201211090800_create_buildpacks.up.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+CREATE TABLE IF NOT EXISTS buildpacks(
+   id uuid DEFAULT uuid_generate_v4 (),
+   namespace VARCHAR (250) NOT NULL,
+   bp_name VARCHAR (250) NOT NULL,
```
Comment (elbandito) [inline db/migrations/20201211090800_create_buildpacks.up.sql]: @jkutner we could probably just call this `name` (without the `bp` prefix)

Comment (jkutner) [reply db/migrations/20201211090800_create_buildpacks.up.sql]: i thought `name` was a reserved word. But I might have been wrong. in any case, we could quote it.

Comment (elbandito) [reply db/migrations/20201211090800_create_buildpacks.up.sql]: Hmm, It might be.  If so, I'm fine leaving it.

File: cmd/index-buildpacks/main.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+func FetchBuildpackConfig(e Entry, imageFn ImageFunction) (Metadata, error) {
+	ref, err := name.ParseReference(e.Address)
+	if err != nil {
+		return Metadata{}, err
+	}
+
+	if _, ok := ref.(name.Digest); !ok {
+
```
Comment (elbandito) [inline cmd/index-buildpacks/main.go]: extra space

File: cmd/index-buildpacks/main.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	for _, s := range m.Stacks {
+		stacks = append(stacks, s.ID)
+	}
+
+	return m, nil
+}
+
+func UpsertMetadata(db *sql.DB, e Entry, m Metadata) error {
+	upsert := `
+insert into buildpacks (namespace, bp_name, version, addr, homepage, description, licenses, stacks)
```
Comment (elbandito) [inline cmd/index-buildpacks/main.go]: I suggest for consistency, we make all the SQL verbiage uppercase.  
```
INSERT INTO buildpacks (namespace, bp_name, version, addr, homepage, description, licenses, stacks) VALUES($1, $2, $3, $4, $5, $6, $7, $8)
ON CONFLICT (namespace, bp_name, version)
DO 
   UPDATE SET homepage = $5, description = $6, licenses = $7, stacks = $8;
```

File: cmd/index-buildpacks/main.go (hunk 3 above)
Comment (jkutner) [reply cmd/index-buildpacks/main.go]: fixed in fabf548

File: cmd/index-buildpacks/main.go (hunk 4 above)
Comment (jkutner) [reply cmd/index-buildpacks/main.go]: fixed in fabf548

File: db/migrations/20201211090800_create_buildpacks.up.sql (hunk 2 above)
Comment (jkutner) [reply db/migrations/20201211090800_create_buildpacks.up.sql]: fixed in 58a2513

**Review by elbandito — APPROVED** [review]

**Pull request merged**